### ISSUE TITLE
[SDL2] FingerDown/Up event is generating also mouse button down/up event 

### ISSUE THREAD

**Issue by Raffaello** [libsdl-org/SDL#11204]
So on a linux Wayland when using a touch screen SDL2 events are generating 2 events:

- mouse button down/up
- finger down/up

and strictly in the order first the mouse button then the finger touch event.

This of course leads to process twice the same unique event.

i've done some research about it and there is already a discussion on it and this issue/patch:
https://github.com/libsdl-org/SDL/issues/755

but that is not document and it looks totally not implemented.

So the only way is to check the event if it is a  `which == SDL_TOUCH_MOUSEID`.

Is there a way to avoid to produce the mouse event for a touch mouse in the first place?

**Comment by Raffaello** [comment]
> Things have changed a lot with SDL since 2013. It sounds like you just need to add this hint?
> 
> `SDL_SetHint(SDL_HINT_TOUCH_MOUSE_EVENTS, "0");`

yep it looks like that.
This hints must be set before init SDL or can be done anytime?

**Comment by slouken** [comment]
> > Things have changed a lot with SDL since 2013. It sounds like you just need to add this hint?
> > `SDL_SetHint(SDL_HINT_TOUCH_MOUSE_EVENTS, "0");`
> 
> yep it looks like that. This hint must be set before init SDL or can be done anytime?

It can be set anytime.

**Comment by AntTheAlchemist** [comment]
You'll probably need to set `SDL_HINT_MOUSE_TOUCH_EVENTS` to "0" as well, if you intend to handle both events separately. It defaults to "0" on desktops, but it'll save confusion if you target other platforms later.

**Comment by Raffaello** [comment]
thanks a lot. that's do it! No more double events when using a touch screen.

`SDL_SetHint(SDL_HINT_TOUCH_MOUSE_EVENTS, "0");`

**Comment by Raffaello** [comment]
Apologies, I am back on this one as i have an update:

This is working on (for what i have tested so far):
- Windows MinGW
- Linux Fedora >= 42

This isn't working on:
- Ubuntu Desktop LTS 24.04 with X11 

Furthermore with Ubuntu Desktop LTS 24.04 with X11 (having 1 mouse) the touch screen is generating an event that has the same mouse instance id of 0 instead of `SDL_TOUCH_MOUSEID` so this is really messing up as it is geneting an event like the physical mouse device has been clicked.

**Comment by icculus** [comment]
My guess here is that X11 itself is making the touches look like mouse input.

**Comment by Kethen** [comment]
Seems to be a thing with XWayland as well on SDL3, when running a test program with `WAYLAND_DISPLAY=""`

The test program can be found at https://github.com/Kethen/sdl_touch_test/

**Comment by slouken** [comment]
Yep, I'm pretty sure that XWayland sends mouse events for touch input without any way of us being able to tell. @Kontrabant can confirm?

**Comment by Kontrabant** [comment]
Hmm, there is a flag to detect emualted pointer events, it just wasn't being used for button events aside from discarding emulated wheel events. The XInput2 backend currently filters all motion events with the emulated flag, so it can probably do the same for button events, as the only case where pointer button events have this flag set is from touch events emulating mouse events, in which case SDL will internally generate the emulated events when necessary, and when coarse scrolling is emulated from scroll valuators, which is handled natively. This is corrorborated by the XInput spec:
```
PointerEmulated signals that the event has been emulated from another
        XI 2.x event for legacy client support, and that this event should
        be ignored if the client listens for these events.  This flag is
        set on scroll ButtonPress and RawButtonPress events (buttons 4, 5, 6
        and 7) if a smooth-scrolling event on the Rel Vert Scroll or
        Rel Horiz Scroll axes was also generated. It is also set on Motion,
        ButtonPress, and ButtonRelease events generated by direct touch devices.
```

@kethen, could you try this branch and see if it fixes your issue? I can't test on a real multitouch device at the moment.

https://github.com/Kontrabant/SDL/tree/xi_drop_button_ev

**Comment by Kethen** [comment]
@Kontrabant  picking the top commit to master (3e68539), it seems to work on the Steamdeck, touching does not generate mouse down anymore in **XWayland**

**Comment by Kontrabant** [comment]
Great, thanks for testing!
